stm32-modules/include/flex-stacker/flex-stacker/tmf8821.hpp

```diff
@@ -0,0 +1,457 @@
+    auto read_register(TOFSensorID sensor_id) -> std::optional<Reg> {
+        using RT = std::optional<Reg>;
+        auto ret = read(sensor_id, Reg::address, 1);
+        if (!ret.has_value()) return RT();
+        auto value = *reinterpret_cast<Reg*>(&ret.value());
+        // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+        return RT(value);
+    }
+
+    template <tmf8821::TMF8821Register Reg>
+    requires WritableRegister<Reg>
+    auto set_register(Reg reg, TOFSensorID sensor_id) -> std::optional<Reg> {
+        using RT = std::optional<Reg>;
+        // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+        auto value = get_register_value(reg);
+        auto ret = write(sensor_id, Reg::address, &value, 1);
+        if (!ret.has_value()) return RT();
+        return RT(reg);
+    }
+
+    /* Makes sure the sensor is ready for communication */
+    auto set_sensor_ready(TOFSensorID sensor_id) -> bool {
+        auto ret = read_register<tmf8821::Enable>(sensor_id);
+        if (!ret.has_value()) {
+            return false;
+        }
+        auto reg = static_cast<tmf8821::Enable>(ret.value());
+        if (!reg.pon || !reg.cpu_ready) {
+            _registers.enable.pon = 1;
+            _registers.enable.powerup_select = reg.powerup_select;
+            update_enable(_registers, sensor_id);
+            // Check if device is ready for comms
+            for (uint8_t i = 0; i < DEFAULT_RETRIES; i++) {
+                // Need to wait after setting Enable reg.
+                _policy->sleep_ms(DEFAULT_SLEEP_MS);
+                ret = read_register<tmf8821::Enable>(sensor_id);
+                if (ret.has_value()) {
+                    reg = static_cast<tmf8821::Enable>(ret.value());
+                    // device is ready for comms
+                    if (reg.pon && reg.cpu_ready) {
+                        return true;
+                    }
+                }
+            }
+            // Device is not ready for comms
+            return false;
+        }
+        return true;
+    }
+
+    auto handle_bootloader(TOFSensorID sensor_id) -> bool {
+        if (tmf8820_image_length < 1 || sensor_id == TOF_NONE) {
+            return false;
+        }
+
+        // ready the driver for image download
+        if (!bl_send_download_init(sensor_id)) {
+            return false;
+        }
+
+        // set the target ram address
+        if (!bl_send_set_address(sensor_id, tmf8820_image_start)) {
+            return false;
+        }
+
+        // write firmware in 128 byte chunks to sensor
+        uint8_t chunk_size = 0;
+        for (uint32_t i = 0; i < tmf8820_image_length; i += BL_DATA_LEN) {
+            chunk_size = (i + BL_DATA_LEN <= tmf8820_image_length)
+                             ? BL_DATA_LEN
+                             : (tmf8820_image_length - i);
+            // send chunk
+            if (!bl_send_image_chunk(sensor_id, (uint8_t*)&tmf8820_image[i],
```

Comment:
imo pointer arithmetic is a bit easier to follow than this but that's a light opinion